services/director-v2/src/simcore_service_director_v2/utils/dags.py

```diff
@@ -1,97 +1,115 @@
     if not node["outputs"]:
         return True
     for output_port in node["outputs"]:
         if output_port is None:
             return True
-    # ok so we have outputs, but maybe the inputs are old? let's check recursively
+    # check if the previous node (if any) are dirty... in which case this one is too
     for input_port in node["inputs"].values():
         if isinstance(input_port, PortLink):
-            if is_node_dirty(full_dag_graph, input_port.node_uuid):
+            if _is_node_dirty(nodes_data_view, input_port.node_uuid):
                 return True
-        else:
-            # FIXME: here we should check if the current inputs are the ones used to generate the current outputs
-            # this could be done by saving the inputs as metadata together with the outputs (see blockchain)
-            # we should compare the current inputs with the inputs used for generating the current outputs!!!
-            pass
-    return False
+    # maybe our inputs changed? let's compute the node hash and compare with the saved one
+    async def get_node_io_payload_cb(node_id: NodeID) -> Dict[str, Any]:
+        return nodes_data_view[str(node_id)]
+
+    computed_hash = await compute_node_hash(node_id, get_node_io_payload_cb)
+    return computed_hash != node["run_hash"]
 
 
 @log_decorator(logger=logger)
-def create_minimal_graph_based_on_selection(
-    full_dag_graph: nx.DiGraph, selected_nodes: List[NodeID]
+def create_complete_dag_graph(workbench: Workbench) -> nx.DiGraph:
+    """creates a complete graph out of the project workbench"""
+    dag_graph = nx.DiGraph()
+    for node_id, node in workbench.items():
+        dag_graph.add_node(
+            node_id,
+            name=node.label,
+            key=node.key,
+            version=node.version,
+            inputs=node.inputs,
+            run_hash=node.run_hash,
+            outputs=node.outputs,
+        )
+        for input_node_id in node.input_nodes:
+            predecessor_node = workbench.get(str(input_node_id))
+            if predecessor_node:
+                dag_graph.add_edge(str(input_node_id), node_id)
+
+    return dag_graph
+
+
+@log_decorator(logger=logger)
+async def create_minimal_computational_graph_based_on_selection(
+    full_dag_graph: nx.DiGraph, selected_nodes: List[NodeID], force_restart: bool
 ) -> nx.DiGraph:
-    selected_nodes_str = [str(n) for n in selected_nodes]
-    # first pass, set the dirty attribute on the graph
+    nodes_data_view: nx.classes.reportviews.NodeDataView = full_dag_graph.nodes.data()
+
+    # first pass, find the nodes that are dirty (outdated)
     for node in nx.topological_sort(full_dag_graph):
-        if node in selected_nodes_str or _node_outdated(full_dag_graph, node):
-            mark_node_dirty(full_dag_graph, node)
+        if _node_computational(nodes_data_view[node]["key"]) and await _node_outdated(
+            nodes_data_view, node
+        ):
+            _mark_node_dirty(nodes_data_view, node)
```

Comment:
Computational nodes are marked/flagged as "dirty" if they are outdated. A node is outdated when 
  - it has no outputs or some of them are missing
  - its inputs are marked as "dirty" (I understand that inputs of this node are the same as outputs of the previous, right?)
  - node run-hash has changed.
  
 I guess this wording comes from git, that refers to a dirty state when a repository has uncommitted changes.  Is that the idea?

Reply:
you don't like dirtiness? 
maybe I can be more consistent. but here was my thinking:
- only checking the inputs is not enough if the node is not computational (which I know is not the case here at the moment).
- so a node is dirty if it is outdated. actually I will change the wording.

Reply:
added some comments in the function